hypershift-operator/controllers/hostedClusterInfrastructure/hostedclusterinfrastructure_controller.go

```diff
@@ -0,0 +1,286 @@
+	}
+
+	// Check for previous work and set Phase=READY if complete
+	infraStatus := meta.IsStatusConditionTrue(hci.Status.Conditions, string(hyperv1.PlatformConfigured))
+	iamStatus := meta.IsStatusConditionTrue(hci.Status.Conditions, string(hyperv1.PlatformIAMConfigured))
+	if infraStatus && iamStatus && hci.DeletionTimestamp.IsZero() {
+		log.Info(fmt.Sprintf("Nothing to do, Hosted Cluster Infrastructure is %s", string(hyperv1.AsExpected)))
+		if hci.Status.Phase != hyperv1.PhaseReady {
+			oHci := hci.DeepCopy()
+			hci.Status.Phase = hyperv1.PhaseReady
+			if err := r.Client.Status().Patch(r.ctx, &hci, client.MergeFrom(oHci)); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Check for a provider secret and exit
+	var providerSecret corev1.Secret
+	var err error
+
+	providerSecretName := hci.Spec.CloudProvider.Name
+	// Nothing to do if the cloudProvider secret name is not present
+	if providerSecretName == "" {
+		log.Info("No provider secret name, can't do a reconcile")
+		r.updateStatusConditionsOnChange(&hci, nil,
+			hyperv1.ProviderSecretConfigured,
+			metav1.ConditionFalse,
+			"No cloudProvider secret reference provided, create a cloudProvider secret in the namespace "+hci.Namespace,
+			hyperv1.NotAsExpected)
+
+		return ctrl.Result{}, nil
+	}
+
+	// Attempt to retreive the cloudProvider secret
+	err = r.Client.Get(ctx, types.NamespacedName{Namespace: hci.Namespace, Name: providerSecretName}, &providerSecret)
+	if err != nil {
+		log.Error(err, fmt.Sprintf("Could not retrieve the provider secret: %s", providerSecretName))
+		return ctrl.Result{RequeueAfter: 30 * time.Second, Requeue: true},
+			r.updateStatusConditionsOnChange(&hci, nil,
+				hyperv1.ProviderSecretConfigured,
+				metav1.ConditionFalse,
+				"The secret "+providerSecretName+" could not be retreived from namespace "+hci.Namespace,
+				hyperv1.NotAsExpected)
+	}
+	if err := r.updateStatusConditionsOnChange(&hci, nil, hyperv1.ProviderSecretConfigured, metav1.ConditionTrue, "Retreived secret "+providerSecretName, string(hyperv1.AsExpected)); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Add a finalizer so the resource can't be deleted till the infrastructure is cleaned up
+	// Also set the Infra-ID if its not set.
+	if !controllerutil.ContainsFinalizer(&hci, DestroyFinalizer) {
+		controllerutil.AddFinalizer(&hci, DestroyFinalizer)
+
+		if hci.Spec.InfraID == "" {
+			hci.Spec.InfraID = hci.Name
+		}
+		if hci.Spec.SubDomain == "" {
+			hci.Spec.SubDomain = hci.Spec.InfraID
+		}
+		if err := r.patchHCI(&hci); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to update infra-id: \"%s\" and error: %w", hci.Spec.InfraID, err)
+		}
+	}
+	log.Info("Using INFRA-ID: " + hci.Spec.InfraID)
+
+	// Destroying Hosted Cluster Infrastructure when scheduled for deletion
+	if !hci.DeletionTimestamp.IsZero() {
+		log.Info("Destroying Infrastructure")
+		return r.destroyAWSInfraIam(&hci, &providerSecret)
+	}
```

Comment:
I worry a bit about how aws specific many of these generic code paths are. For example, here we directly call this `r.destroyAWSInfraIam` function for the generic "destroy" logic. And then that AWS specific function is the function that removes the finalizer from the HCI object.

Little nuances like this make it difficult to plug in additional providers in the future. It's not a huge deal, but it means the next person working in on this will have the job of refactoring much of this controller to be provider agnostic.

Reply:
For now there is only:
* AWS
* Azure
* PowerVS.

I was kind of expecting to create an Interface group like this for each provider(AWS & Azure are there). This way the testing can be stubbed out. The idea with the interface was not so much to generalize as it was to be able to avoid the actual provider calls for FVT.

Let me move to a general entry and switch between the infra interfaces.